File: go/prime_factor/prime.go
package prime_factor

// listPrimes should return all prime numbers p, (p < n) *in linear time*.
// For example, for n = 26, the answer should be
// [2, 3, 5, 7, 11, 13, 17, 19, 23]
func listPrimes(n int) []int {
	factors := make([]int16, n)

	var primes []int
	for i := 2; i < n; i++ {
		f := int(factors[i])
		if f == 0 {
			f = i
			primes = append(primes, i)
		}

		for _, p := range primes {
			if p > f || p*i >= n {
				break
			}

			factors[p*i] = int16(p)
		}
	}

	return primes
}

File: python/prime_factor/prime.py

# list_primes should return all prime numbers p, (p < n) *in linear time*.
# For example, for n = 26, the answer should be
# [2, 3, 5, 7, 11, 13, 17, 19, 23]
def list_primes(n: int) -> list[int]:
	# raise NotImplementedError('TODO')

	factors = [0]*n
	primes = []

	for i in range(2, n):
		f = factors[i]
		if f == 0:
			f = i
			primes.append(i)

		for p in primes:
			if p > f or p*i >= n:
				break

			factors[p*i] = p

	return primes
